zcash_primitives/src/extensions/transparent.rs

```diff
@@ -0,0 +1,199 @@
+/// is treated as opaque to all but extension corresponding to the
+/// encapsulated extension_id value.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Witness {
+    pub extension_id: u32,
+    pub mode: u32,
+    pub payload: Vec<u8>,
+}
+
+impl Witness {
+    /// Produce the intermediate format for an extension-specific witness
+    /// type.
+    pub fn from<P: ToPayload>(extension_id: u32, value: &P) -> Witness {
+        let (mode, payload) = value.to_payload();
+        Witness {
+            extension_id,
+            mode,
+            payload,
+        }
+    }
+
+    /// Attempt to parse an extension-specific witness value from the
+    /// intermediate representation.
+    pub fn try_to<P: FromPayload>(&self) -> Result<P, P::Error> {
+        P::from_payload(self.mode, &self.payload)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error<E> {
+    InvalidExtensionId(u32),
+    ProgramError(E),
+}
+
+impl<E: fmt::Display> fmt::Display for Error<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidExtensionId(extension_id) => {
+                write!(f, "Unrecognized program type id {}", extension_id)
+            }
+
+            Error::ProgramError(err) => write!(f, "Program error: {}", err),
+        }
+    }
+}
+
+/// This is the primary trait which must be implemented by an extension type
+/// for that type to be eligible for inclusion in Zcash consensus rules.
+pub trait Extension<C> {
+    /// Extension-specific precondition type. The extension will need to implement
+    /// [`FromPayload<Error = Self::Error>`] for this type in order for their
+    /// extension to be eligible for integration into consensus rules.
+    type P;
+
+    /// Extension-specific witness type. The extension will need to implement
+    /// [`FromPayload<Error = Self::Error>`] for this type in order for their
+    /// extension to be eligible for integration into consensus rules.
+    type W;
+
+    /// Extension-specific error type. This should encompass both parsing and verification errors.
+    type Error;
+
+    /// This is the primary method that an extension must implement. Implementations should
+    /// return [`Ok(())`] if verification of the witness succeeds against the supplied
+    /// precondition, and an error in any other case.
+    fn verify_inner(
+        &self,
+        precondition: &Self::P,
+        witness: &Self::W,
+        context: &C,
+    ) -> Result<(), Self::Error>;
+
+    /// This is a convenience method intended for use by consensus nodes at the integration
+    /// point to provide easy interoperation with the opaque, cross-extension
+    /// `Precondition` and `Witness` types.
+    fn verify(
+        &self,
+        precondition: &Precondition,
+        witness: &Witness,
+        context: &C,
+    ) -> Result<(), Self::Error>
+    where
+        Self::P: FromPayload<Error = Self::Error>,
```

Comment:
Can these bounds be placed on the associated type in the declaration of `Extension` to avoid having to write `where` bounds for this in various places?

Reply:
I thought about this. Here's my reason for not placing these constraints at the declaration site, based on my experience with type polymorphism in other languages:

Basically, if you place constraints at the use site instead of the declaration site, you provide more flexibility to the implementer; they could choose to implement an Extension where those constraints aren't observed, then provide some alternate mechanism for use at the declaration site. Basically, you defer the decision of what `P` has to be; the caller could choose not to use this convenience `verify` if they had already performed parsing by some other mechanism, for example. 

This is a pretty common pattern where type parameters appear in contravariant position in functional languages, because it permits the implementation of methods like ```contramap<Q>(&self, f: impl FnOnce(Q) -> P) -> impl Extension<Precondition = Q>``` If `P` (nee `Precondition`) is constrained at the declaration site, you can't do this.